README.md
# appsflyer


#### Documentation

[Documentation](https://support.appsflyer.com/hc/en-us)


#### [Pull API](https://support.appsflyer.com/hc/en-us/articles/207034346-Pull-APIs-Pulling-AppsFlyer-Reports-by-APIs)

 - [getting Raw Report Data](https://support.appsflyer.com/hc/en-us/articles/208387843-Raw-Data-Reports-V5-)
  
  Example:
  
  ```php
  $api = new \appsflyer\api\PullApi\RawDataReport($apiKey);
  try {
      $api->setFilters(['from' => '2017-02-27', 'to' => '2017-03-03'])
          ->addBeforeRunCallback(function ($params) {
              var_dump($params);
          })
          ->getInAppEventsReport($idApp);
 
      $arr = $api->getCsvAsArray();
      //to get mapped array
      $mapped = $api->mapCsvArrayToColumns($arr);
      var_dump($api->getLastCurlError(), $api->getLastCurlFile());
  } catch (\Exception $e) {
      var_dump($e->getMessage());
  }
  ```

CHANGELOG.md
# 1.0.4
 - Array of Pull API Raw data report columns
 - method for getting mapped raw data report as array

# 1.0.0
 - Pull API get raw report data method

src/appsflyer/api/BaseApi.php
<?php
/**
 * @author Sergey Kubrey <kubrey.work@gmail.com>
 *
 */

namespace appsflyer\api;


class BaseApi
{
    protected $lastCurlInfo = [];
    protected $lastCurlError;
    protected $lastCurlResponse;
    protected $lastCurlOptions;
    protected $lastCurlFile = null;
    protected $tmpFolder;

    protected $curl;

    protected $curlOptions;
    protected $tmpFilePointer;
    protected $tmpFileName;

    /**
     * @param $url
     * @return null|string path to file
     * @throws \Exception
     */
    public function prepareCurlFileRequest($url) {
        $this->resetCurl();
        if (!$this->tmpFileName) {
            throw  new \Exception("No temporary file name is set");
        }
        $this->curl = curl_init();
        $destination = dirname(__FILE__) . "/../../tmp/" . $this->tmpFileName;
        $this->tmpFilePointer = fopen($destination, 'w+');
        $options = [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HEADER => false,
            CURLOPT_CONNECTTIMEOUT => 30,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_SSL_VERIFYPEER => false,
            CURLOPT_URL => $url,
            CURLOPT_FILE => $this->tmpFilePointer
        ];
        $this->curlOptions = $options;
        $this->lastCurlFile = $destination;

        return $this;
    }

    /**
     * @param array $options
     * @return $this
     */
    public function setAdditionalCurlOptions($options = []) {
        if ($options && is_array($options)) {
            foreach ($options as $key => $val) {
                $this->curlOptions[$key] = $val;
            }
        }

        return $this;
    }

    /**
     * @return $this
     * @throws \Exception
     */
    protected function runCurl() {

        curl_setopt_array($this->curl, $this->curlOptions);
        $this->lastCurlResponse = curl_exec($this->curl);
        $this->lastCurlError = curl_error($this->curl);
        $this->lastCurlInfo = curl_getinfo($this->curl);
        curl_close($this->curl);
        if (is_resource($this->tmpFilePointer)) {
            fclose($this->tmpFilePointer);
        }
        if ($this->lastCurlError) {
            throw  new \Exception($this->lastCurlError);
        }
        if (!isset($this->lastCurlInfo['http_code']) || $this->lastCurlInfo['http_code'] != "200") {
            throw new \Exception("Curl Http Code is " . (isset($this->lastCurlInfo['http_code']) ?
                    $this->lastCurlInfo['http_code'] : "empty"));
        }

        return $this;
    }

    /**
     * @return $this
     */
    public function resetCurl() {
        if ($this->lastCurlFile) {
            @unlink($this->lastCurlFile);
        }
        $this->lastCurlInfo = [];
        $this->lastCurlResponse = null;
        $this->lastCurlError = null;
        $this->lastCurlFile = null;
        return $this;
    }

    public function getLastCurlFile() {
        return $this->lastCurlFile;
    }

    public function setLastCurlFile($path) {
        if (!is_file($path) || !is_readable($path)) {
            throw  new \Exception("Path is not valid or readable");
        }
        $this->lastCurlFile = $path;
        return $this;
    }

    public function getLastCurlInfo() {
        return $this->lastCurlInfo;
    }

    public function getLastCurlError() {
        return $this->lastCurlError;
    }

    public function getLastCurlResponse() {
        return $this->lastCurlResponse;
    }

    /**
     * @param $name
     * @return $this
     */
    public function setTmpFileName($name) {
        $this->tmpFileName = $name;
        return $this;
    }

    public function __destruct() {
        if (is_file($this->lastCurlFile)) {
            unlink($this->lastCurlFile);
        }
    }

}

examples/parse_raw_data_csv.php
<?php
/**
 * @author Sergey Kubrey <kubrey.work@gmail.com>
 *
 */

$apiKey = isset($argv[1]) ? $argv[1] : null;
if(!$apiKey){
    echo "No api key\n";
    exit();
}

require __DIR__ . "/../src/appsflyer/api/BaseApi.php";
require __DIR__ . "/../src/appsflyer/api/PullApi/BasePullApi.php";
require __DIR__ . "/../src/appsflyer/api/PullApi/RawDataReport.php";

$api = new \appsflyer\api\PullApi\RawDataReport($apiKey);
$file = __DIR__.'/../src/tmp/id1151406354_non_organic_in_app_events_2017-02-27_2017-03-03_IZLPA.csv';

try {
    $api->setLastCurlFile($file);
    $api->setTmpFileName("testname");

    $arr = $api->getCsvAsArray();
    $map = $api->mapCsvArrayToColumns($arr);
    var_dump($map);
    //    var_dump($arr);
    //    $api->resetCurl();
//    var_dump($api->getLastCurlError(), $api->getLastCurlFile());
} catch (\Exception $e) {
    var_dump($api->getLastCurlResponse());
    var_dump($e->getMessage());
}

src/appsflyer/api/PullApi/BasePullApi.php
<?php
/**
 * @author Sergey Kubrey <kubrey.work@gmail.com>
 *
 */

namespace appsflyer\api\PullApi;

use appsflyer\api\BaseApi;

/**
 * Class BasePullApi
 * @package appsflyer\api\PullApi
 * Limits: Max line count: 200K | Oldest available data: 90 Days ago. Max requests 12 per day per app, 60 per account in total.
 */
class BasePullApi extends BaseApi
{
    protected $apiKey = null;
    protected $version = 5;
    protected $baseUrl = "hq.appsflyer.com/export";
    protected $scheme = "https";
    protected $filters = [];
    protected $action;
    protected $applicationId;

    protected $resultUrl;
    protected $response;

    protected $afterRunCallbacks = [];
    protected $beforeRunCallbacks = [];

    protected $method = null;

    const METHOD_GET = 'GET';
    const METHOD_POST = 'POST';
    const METHOD_FILE = 'FILE';
    const METHOD_PUT = 'PUT';

    public function __construct($apiKey) {
        $this->apiKey = $apiKey;
    }

    public function setVersion($version) {
        $this->version = $version;
        return $this;
    }

    public function setFilters($filters = []) {
        if (!is_array($filters)) {
            throw  new \Exception("Filters should be an array");
        }
        $this->filters = $filters;
        return $this;
    }

    public function setAction($action) {
        if (!$action) {
            throw new \Exception("No action");
        }
        $this->action = trim($action, " /");
        return $this;
    }

    public function setApplicationId($applicationId) {
        if (!$applicationId) {
            throw new \Exception("No application id");
        }
        $this->applicationId = $applicationId;
        return $this;
    }

    /**
     * @return $this
     * @throws \Exception
     */
    public function run() {
        $fn = null;
        switch ($this->method) {
            case static::METHOD_FILE:
                $fn = "prepareCurlFileRequest";
                break;
            case static::METHOD_GET:

                break;
        }
        if (!$fn && !method_exists($this, $fn)) {
            throw new \Exception("Invalid method is set");
        }

        $this->composeResultUrl()
            ->callBeforeRunCallbacks()
            ->$fn($this->resultUrl)
            ->runCurl()
            ->callAfterRunCallbacks();

        return $this;
    }

    /**
     * @return $this
     * @throws \Exception
     */
    protected function composeResultUrl() {
        $url[] = $this->scheme . "://" . $this->baseUrl;
        $url[] = $this->applicationId;
        $url[] = $this->action;
        $url[] = "v" . $this->version;

        $queryParts = [];
        $queryParts['api_token'] = $this->apiKey;
        foreach ($this->filters as $key => $filterVal) {
            $queryParts[$key] = $filterVal;
        }

        foreach ($url as $urlPart) {
            if (!$urlPart) {
                throw  new \Exception("Url is not filled properly");
            }
        }
        $this->baseUrl = implode(DIRECTORY_SEPARATOR, $url);
        $getParams = http_build_query($queryParts);
        $this->resultUrl =$this->baseUrl."?" . $getParams;

        return $this;
    }

    /**
     * @param $callback
     * @return $this
     * @throws \Exception
     */
    public function addBeforeRunCallback($callback) {
        if (!is_callable($callback)) {
            throw  new \Exception("Before run callback is invalid");
        }
        $this->beforeRunCallbacks[] = $callback;
        return $this;
    }

    /**
     * @param $callback
     * @return $this
     * @throws \Exception
     */
    public function addAfterRunCallback($callback) {
        if (!is_callable($callback)) {
            throw  new \Exception("After run callback is invalid");
        }
        $this->afterRunCallbacks[] = $callback;
        return $this;
    }

    /**
     * @return $this
     */
    public function callBeforeRunCallbacks() {
        foreach ($this->beforeRunCallbacks as $cb) {
            if (!is_callable($cb)) {
                continue;
            }
            $params = [
                'url' => $this->resultUrl,
                'time' => time(),
                'timezone' => date_default_timezone_get(),
                'action' => $this->action
            ];
            call_user_func_array($cb, ['params' => $params]);
        }

        return $this;
    }

    /**
     * @return $this
     */
    public function callAfterRunCallbacks() {
        foreach ($this->afterRunCallbacks as $cb) {
            if (!is_callable($cb)) {
                continue;
            }
            $params = [
                'url' => $this->resultUrl,
                'time' => time(),
                'timezone' => date_default_timezone_get(),
                'action' => $this->action,
                'curl' => [
                    'error' => $this->lastCurlError,
                    'info' => $this->lastCurlInfo,
                    'file' => $this->lastCurlFile
                ]
            ];
            call_user_func_array($cb, ['params' => $params]);
        }

        return $this;
    }

    public function setPostMethod() {
        $this->method = static::METHOD_POST;
        return $this;
    }

    public function setGetMethod() {
        $this->method = static::METHOD_GET;
        return $this;
    }

    /**
     * Get request and downloading file
     * @return $this
     */
    public function setFileMethod() {
        $this->method = static::METHOD_FILE;
        return $this;
    }


}

src/appsflyer/api/PullApi/RawDataReport.php
<?php
/**
 * @author Sergey Kubrey <kubrey.work@gmail.com>
 *
 */

namespace appsflyer\api\PullApi;

/**
 * @see https://support.appsflyer.com/hc/en-us/articles/208387843-Raw-Data-Reports-V5-
 * Class RawDataReport
 * @package appsflyer\api\PullApi
 */
class RawDataReport extends BasePullApi
{
    const IN_APP_EVENTS_REPORT = 'in_app_events_report';

    public function getInAppEventsReport($applicationId) {
        $this->setApplicationId($applicationId)
            ->setAction(static::IN_APP_EVENTS_REPORT)
            ->setFileMethod()
            ->setTmpFileName(static::IN_APP_EVENTS_REPORT . "_" . uniqid() . ".csv")
            ->run();
    }

    /**
     * @return array
     * @throws \Exception
     */
    public function getCsvAsArray() {
        if (!$this->tmpFileName || !$this->lastCurlFile) {
            throw new \Exception("No file");
        }

        $csvRows = [];
        if (($handle = fopen($this->lastCurlFile, "r")) !== false) {
            while (($row = fgetcsv($handle, 0, ",")) !== false) {
                $csvRows[] = $row;
            }
            fclose($handle);
        }
        return $csvRows;
    }

    /**
     * @param $csvArray
     * @return array
     */
    public function mapCsvArrayToColumns($csvArray) {
        if (!$csvArray) {
            return [];
        }
        $columns = static::getColumns();
        $headRow = $csvArray[0];
        $mappedArray = [];
        foreach ($csvArray as $index => $row) {
            $mappedRow = [];
            if ($index == 0) {
                //passing through titles
                continue;
            }
            foreach ($row as $cellIndex => $value) {
                $columnTitle = $headRow[$cellIndex];
                if (!$columnTitle) {
                    //this should not happen
                    continue;
                }
                $innerTitle = array_search($columnTitle, $columns);
                if (!$innerTitle) {
                    //this should not happen either
                    continue;
                }
                $mappedRow[$innerTitle] = $value;

            }
            $mappedArray[] = $mappedRow;
        }
        return $mappedArray;
    }

    /**
     * @param int $version
     * @return array
     */
    public static function getColumns($version = 5) {
        $columns['v5'] = [
            'attributed_touch_type' => 'Attributed Touch Type',
            'attributed_touch_time' => 'Attributed Touch Time',
            'install_time' => 'Install Time',
            'af_prt' => 'Partner',
            'media_source' => 'Media Source',
            'is_retargeting' => 'Is Retargeting',
            'retargeting_conversion_type' => 'Retargeting Conversion Type',
            'af_channel' => 'Channel',
            'af_keywords' => 'Keywords',
            'campaign' => 'Campaign',
            'af_c_id' => 'Campaign ID',
            'af_adset' => 'Adset',
            'af_adset_id' => 'Adset ID',
            'af_ad' => 'Ad',
            'af_ad_id' => 'Ad ID',
            'af_ad_type' => 'Ad Type',
            'af_site_id' => 'Site ID',
            'af_sub1' => 'Sub Param 1',
            'af_sub2' => 'Sub Param 2',
            'af_sub3' => 'Sub Param 3',
            'af_sub4' => 'Sub Param 4',
            'af_sub5' => 'Sub Param 5',
            'http_referrer' => 'HTTP Referrer',
            'original_url' => 'Original URL',
            'user_agent' => 'User Agent',
            'af_cost model' => 'Cost Model',
            'af_cost_value' => 'Cost Value',
            'af_cost currency' => 'Cost Currency',
            'contributor1_af_prt' => 'Contributor 1 Partner',
            'contributor1_media_source' => 'Contributor 1 Media Source',
            'contributor1_campaign' => 'Contributor 1 Campaign',
            'contributor1_touch_type' => 'Contributor 1 Touch Type',
            'contributor1_touch_time' => 'Contributor 1 Touch Time',
            'contributor2_af_prt' => 'Contributor 2 Partner',
            'contributor2_media_source' => 'Contributor 2 Media Source',
            'contributor2_campaign' => 'Contributor 2 Campaign',
            'contributor2_touch_type' => 'Contributor 2 Touch Type',
            'contributor2_touch_time' => 'Contributor 2 Touch Time',
            'contributor3_af_prt' => 'Contributor 3 Partner',
            'contributor3_media_source' => 'Contributor 3 Media Source',
            'contributor3_campaign' => 'Contributor 3 Campaign',
            'contributor3_touch_type' => 'Contributor 3 Touch Type',
            'contributor3_touch_time' => 'Contributor 3 Touch Time',
            'country_code' => 'Country Code',
            'ip' => 'IP',
            'region' => 'Region',
            'state' => 'State',
            'city' => 'City',
            'wifi' => 'WIFI',
            'operator' => 'Operator',
            'carrier' => 'Carrier',
            'language' => 'Language',
            'appsflyer_id' => 'AppsFlyer ID',
            'advertising_id' => 'Advertising ID',
            'idfa' => 'IDFA',
            'android_id' => 'Android ID',
            'customer_user_id' => 'Customer User ID',
            'imei' => "IMEI",
            'idfv' => 'IDFV',
            'platform' => 'Platform',
            'device_type' => 'Device Type',
            'os_version' => 'OS Version',
            'app_version' => 'App Version',
            'sdk_version' => 'SDK Version',
            'app_id' => 'App ID',
            'app_name' => 'App Name',
            'bundle_id' => 'Bundle ID',
            'af_attribution_lookback' => 'Attribution Lookback',
            'event_time' => 'Event Time',
            'event_name' => 'Event Name',
            'event_value' => 'Event Value',
            'event_revenue' => 'Event Revenue',
            'event_revenue_usd' => 'Event Revenue USD',
            'event_revenue_currency' => 'Event Revenue Currency',
            'is_receipt_validated' => 'Is Receipt Validated',
            'postal_code' => 'Postal Code',
            'dma' => 'DMA',
            'af_reengagement_window' => 'Reengagement window',
            'event_source' => 'Event source',
            'is_primary_attribution' => 'Is Primary Attribution',
            'af_sub_site_id' => 'Sub Site ID'
        ];

        return isset($columns['v' . $version]) ? $columns['v' . $version] : [];
    }
}

examples/api_pull_raw_data_report.php
<?php
/**
 * @author Sergey Kubrey <kubrey.work@gmail.com>
 * Set api key as cli argument
 * eg >php api_pull_raw_data_report.php api_key
 */
$apiKey = isset($argv[1]) ? $argv[1] : null;
if(!$apiKey){
    echo "No api key\n";
    exit();
}
var_dump($apiKey);
require __DIR__ . "/../src/appsflyer/api/BaseApi.php";
require __DIR__ . "/../src/appsflyer/api/PullApi/BasePullApi.php";
require __DIR__ . "/../src/appsflyer/api/PullApi/RawDataReport.php";

$api = new \appsflyer\api\PullApi\RawDataReport($apiKey);
try {
    $api->setFilters(['from' => '2017-02-27', 'to' => '2017-03-03'])
        ->addBeforeRunCallback(function ($params) {
            var_dump($params);
        })
        ->getInAppEventsReport("id1151406354");

    $arr = $api->returnAsArray();
//    var_dump($arr);
//    $api->resetCurl();
    var_dump($api->getLastCurlError(), $api->getLastCurlFile());
} catch (\Exception $e) {
    var_dump($api->getLastCurlResponse());
    var_dump($e->getMessage());
}

